deadlock-dotnet-sdk/Domain/NativeMethods.cs

```diff
@@ -1,157 +1,384 @@
+        }
 
-            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = CchRmMaxAppName + 1)]
-            private readonly string strAppName;
+        SYSTEM_HANDLE_INFORMATION_EX retVal = *pSysInfoBuffer;
 
-            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = CchRmMaxSvcName + 1)]
-            private readonly string strServiceShortName;
+        retVal.CheckAccess();
 
-            private readonly RmAppType ApplicationType;
-            private readonly uint AppStatus;
-            private readonly uint TSSessionId;
-            [MarshalAs(UnmanagedType.Bool)] private readonly bool bRestartable;
-        }
+        Marshal.FreeHGlobal((IntPtr)pSysInfoBuffer);
+        Marshal.FreeHGlobal((IntPtr)returnLength);
 
-        #endregion
+        return retVal.AsSpan();
+    }
+
+    #endregion Methods
 
-        #region DllImport
+    #region Structs
 
-        [DllImport("rstrtmgr.dll", CharSet = CharSet.Unicode)]
-        private static extern int RmRegisterResources(uint pSessionHandle, uint nFiles, string[] rgsFilenames,
-            uint nApplications, [In] RmUniqueProcess[] rgApplications, uint nServices, string[] rgsServiceNames);
+    /// <summary>
+    /// The <see href="https://www.geoffchappell.com/studies/windows/km/ntoskrnl/api/ex/sysinfo/handle_ex.htm"><c>SYSTEM_HANDLE_INFORMATION_EX</c></see>
+    /// struct is 0x24 or 0x38 bytes in 32-bit and 64-bit Windows, respectively. However, Handles is a variable-length array.
+    /// </summary>
+    public unsafe readonly struct SYSTEM_HANDLE_INFORMATION_EX
+    {
+#pragma warning disable CS0649
+
+        /// <summary>
+        /// As documented unofficially, NumberOfHandles is a 4-byte or 8-byte ULONG_PTR in 32-bit and 64-bit Windows, respectively.<br/>
+        /// This is not to be confused with uint* or ulong*.
+        /// </summary>
+        public readonly UIntPtr NumberOfHandles;
+        public readonly UIntPtr Reserved;
+        public readonly SYSTEM_HANDLE_TABLE_ENTRY_INFO_EX* Handles;
 
-        [DllImport("rstrtmgr.dll", CharSet = CharSet.Unicode)]
-        private static extern int RmStartSession(out uint pSessionHandle, int dwSessionFlags, string strSessionKey);
+#pragma warning restore CS0649
 
-        [DllImport("rstrtmgr.dll")]
-        private static extern int RmEndSession(uint pSessionHandle);
+        public Span<SYSTEM_HANDLE_TABLE_ENTRY_INFO_EX> AsSpan() => new(Handles, (int)NumberOfHandles);
 
-        [DllImport("rstrtmgr.dll")]
-        private static extern int RmGetList(uint dwSessionHandle, out uint pnProcInfoNeeded, ref uint pnProcInfo,
-            [In, Out] RmProcessInfo[] rgAffectedApps, ref uint lpdwRebootReasons);
+        /// <summary>
+        /// Test for memory access. System.AccessViolationException due to these values being in a protected memory range is a problem.
+        /// </summary>
+        internal void CheckAccess()
+        {
+            Console.WriteLine(Handles[0].HandleValue);
```

Comment:
Any and all R/W operations on the system handles' memory range causes a `System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.`

I've already confirmed this is unrelated to the "JustMyCode" debug setting.

Reply:
If I can't find the reason for this within the next few days, I'll have to put this Pull Request on hold and come back to it later.

Reply:
I have bad news. After I picked this up again after an unplanned break, I did some troubleshooting to find a solution to the AccessViolationException when accessing SYSTEM_HANDLE_TABLE_ENTRY objects' fields. It didn't matter if the host process was in Debug mode and had been run as Administrator. The exception somehow breaks out of a try-catch statement which makes it even more vexing.

I tried taking a look at System Informer's (formerly Process Hacker) relevant source code, but nothing stood out to me.

Tomorrow, I'll try looking at other apps and libraries for examples of what *should* be done.

Reply:
I finally managed to read the raw system handles and copy them to a managed array.
Replacing the `SYSTEM_HANDLE_TABLE_ENTRY_INFO_EX*` field with a non-pointer `SYSTEM_HANDLE_TABLE_ENTRY_INFO_EX`, then pinning the field's object, and finally instantiating a ReadOnlySpan from field's address and the value of NumberOfHandles.

Two months...well, less than that in total work put in, but it took way too long to find a solution.

Reply:
I've finished implementing System Informer's alternative method for identifying a handles object type. It works wonderfully. I only experienced a couple of hiccups during development, too. Firstly, misreading the summary of a Marshal method's overload, causing it to read twice as many bytes as expected. And secondly, wondering why a method was failing to properly iterate through memory after I forgot to implement a ported memory-reading function.

So far, it can get the system handles and identify the Type of the handle's object. Sometimes, it can even get file and process information (but reading another process's memory for its PEB is a W.I.P.)

The debug build runs slow, however. Very slow. It once threw more than 1,000 exceptions in under 10 seconds. Most of them were due to failures to access a process's PEB.

The changes have been pushed to my handle-cli testing branch. 
https://github.com/BinToss/deadlock-dotnet-sdk/compare/unsafeCloseHandle...BinToss:deadlock-dotnet-sdk:handle-cli